packages/falso/src/lib/core/core.ts

```diff
@@ -11,21 +12,83 @@ type Return<T, O extends FakeOptions> = [O] extends [never]
   : T;
 
 export function fake<T, Options extends FakeOptions>(
-  data: T[] | ((i: number) => T),
+  data: T[] | (() => T),
   options?: Options
 ): Return<T, Options> {
-  const dataSource = Array.isArray(data) ? () => randElement(data) : data;
+  if (Array.isArray(data)) {
+    return fakeFromArray(data, options) as any;
+  }
+
+  return fakeFromFunction(data, options) as any;
+}
+
+export function fakeFromFunction<T, Options extends FakeOptions>(
+  data: () => T,
+  options?: Options
+) {
+  if (!options?.length) {
+    return data();
+  }
+
+  const priority = options?.priority ?? 'length';
+
+  if (priority === 'length') {
+    return Array.from({ length: options.length }, (_, index) => data());
+  }
+
+  const items: T[] = [];
+
+  let attempts = 0;
+  const maxAttempts = options.length * 2;
+
+  while (items.length < options.length) {
+    if (attempts >= maxAttempts) {
+      break;
+    }
 
+    const item = data();
+
+    if (!items.includes(item)) {
+      items.push(item);
+    }
+
+    attempts++;
+  }
+
+  return items;
+}
+
+export function fakeFromArray<T, Options extends FakeOptions>(
+  data: T[],
+  options?: Options
+) {
   if (!options?.length) {
-    return dataSource(0) as any;
+    return randElement(data);
   }
 
+  const priority = options?.priority ?? 'length';
+
+  if (priority === 'length') {
+    return Array.from({ length: options.length }, (_, index) =>
+      randElement(data)
+    );
+  }
+
+  const clonedData: T[] = JSON.parse(JSON.stringify(data));
   return Array.from({ length: options.length }, (_, index) =>
-    dataSource(index)
-  ) as any;
+    randUniqueElement(clonedData)
+  ).filter((item) => item);
+}
+
+export function randUniqueElement<T>(arr: T[]): T {
+  const index = Math.floor(random() * arr.length);
+  const item = arr[index];
+  arr.splice(index, 1);
```

Comment:
This is an exported function and I'm not sure that it should have side effects (I don't think that anyone might expect that).
Maybe this should be managed in fakeFromArray